### DISCUSSION TITLE
Client Certificate Authentication

### DISCUSSION THREAD

**Discussion by CICADA-dev64bit** [Kaliumhexacyanoferrat/GenHTTP#548] [Q&A] [answered]
we want to achieve client certificate authentication(mTLS), but we are not finding any APIs in the documentation
1. Is Client Certificate Authentication supported by GenHTTP?
2. Do we need to change the engine to Kestrel to achieve Client certificate authentication? If yes, can you please provide some snippets

**Comment by Kaliumhexacyanoferrat** [comment]
Client certfificate authentication is currently not possible via the API, neither in Kestrel nor in the internal engine. As this is not too complicated (allow to somehow set the `RemoteCertificateValidationCallback` on the `SslServerAuthenticationOptions` of the `SslStream`), I will create an issue for this and implement it in one of the next releases. Thanks for your input.

**Reply by CICADA-dev64bit** [reply]
Hey, can you provide some snippets on how we can achieve mTLS.

Thanks in advance

**Reply by Kaliumhexacyanoferrat** [reply]
As this is not implemented yet I cannot. Will be available in one of the next releases.

**Reply by CICADA-dev64bit** [reply]
okay thanks

**Reply by Kaliumhexacyanoferrat** [reply]
See https://github.com/Kaliumhexacyanoferrat/GenHTTP/issues/556

**Reply by Kaliumhexacyanoferrat** [reply, marked as answer]
@CICADA-dev64bit To give an update: There is a merge request that allows client authentication via API. This will require some polishing and testing, but I think we can release this feature during the next week.

**Reply by CICADA-dev64bit** [reply]
Hey @Kaliumhexacyanoferrat, thanks for updating me with the enhancements, waiting for the next release. Although I did some changes to support the mTLS  and it was working as expected, if you are on it then I wont create PR.

Thanks for getting back so quick with the change

**Reply by Kaliumhexacyanoferrat** [reply]
This is great news - feel free to review my pull request to check whether it contains all the features you need and whether there is anything that can be solved in a more elegant way.

**Reply by Kaliumhexacyanoferrat** [reply]
Client Certificates are available in GenHTTP 9.1, see [the documentation](https://genhttp.org/documentation/server/security/#client-certificates). There is also a [new concern](https://genhttp.org/documentation/content/concerns/authentication/#client-certificate-authentication) to authenticate on API level.
